### PR TITLE
DRMAAV2-18: Interfaces as Defined in specification

### PR DESCRIPTION
**Pull request by vinodchitrali**
#### Issue-ID

* **[DRMAAV2-18](https://pbspro.atlassian.net/browse/DRMAAV2-18)**

#### Problem description
* Interfaces as defined in DRMAAv2 specification 

#### Cause / Analysis
* N/A

#### Solution description
* Stub definition of all the interfaces defined OGF in DRMAAv2 specification 

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: api/c-binding/drmaa2.h [hunk 1]
```diff
@@ -38,7 +38,427 @@
 #ifndef DRMAA2_H_
```
Comment (mike0042) [inline api/c-binding/drmaa2.h]: No copyright header?

File: api/cpp-binding/drmaa2.hpp [hunk 2]
```diff
@@ -38,8 +38,1236 @@
 #ifndef DRMAA2_HPP_
```
Comment (mike0042) [inline api/cpp-binding/drmaa2.hpp]: No copyright header?

File: api/c-binding/drmaa2.cpp [hunk 3]
```diff
@@ -40,9 +40,1647 @@
```
Comment (mike0042) [inline api/c-binding/drmaa2.cpp]: No copyright header?

File: api/c-binding/drmaa2.h (hunk 1 above)
Comment (vinodchitrali) [inline api/c-binding/drmaa2.h]: There is copyright header. It was there for line 1 to 38. It was committed earlier

Comment (mike0042) [inline api/c-binding/drmaa2.h]: So there is. My mistake.

**Comment by mike0042** [comment]
Is the code building at this point?

File: api/c-binding/drmaa2.h [hunk 4]
```diff
@@ -38,7 +38,427 @@
...
+	DRMAA2_UNSET_LISTTYPE = -1,
+	DRMAA2_STRINGLIST = 0,
+	DRMAA2_JOBLIST = 1,
+	DRMAA2_QUEUEINFOLIST = 2,
+	DRMAA2_MACHINEINFOLIST = 3,
+	DRMAA2_SLOTINFOLIST = 4,
+	DRMAA2_RESERVATIONLIST = 5
+} drmaa2_listtype;
+typedef void (*drmaa2_list_entryfree)(void **value);
+void drmaa2_string_list_default_callback(void **value);
```
Comment (arungrover) [inline api/c-binding/drmaa2.h]: Just for readability can we please have a line spaced between things that are logically not aligned. Just like it is there in the spec

File: api/c-binding/drmaa2.cpp [hunk 5]
```diff
@@ -40,9 +40,1647 @@
...
+ *  @brief  This function deletes an element from the list
+ *
+ *  @param[in]	l - Pointer to drmaa list
+ *  @param[in]	pos - position of the element which needs to be deleted
+
+ *  @return drmaa2_error - DRMAA2_INVALID_ARGUMENT
+ *						 - DRMAA2_SUCCESS
+ *
+ */
+drmaa2_error drmaa2_list_del(drmaa2_list l, long pos) {
```
Comment (arungrover) [inline api/c-binding/drmaa2.cpp]: Declaration says that it is a "const long pos"

File: api/c-binding/drmaa2.cpp [hunk 6]
```diff
@@ -40,9 +40,1647 @@
...
+ *
+ *  @param[in]	l - Pointer to drmaa list
+ *  @param[in]	index - postion from whcih the data has to be taken
+
+ *  @return
+ *  	void* - pointer to the element retured as void*
+ *  	NULL  - If the position is not valid
+ *
+ */
+const void *drmaa2_list_get(const drmaa2_list l, long index) {
```
Comment (arungrover) [inline api/c-binding/drmaa2.cpp]: it needs to be const long index

File: api/c-binding/drmaa2.cpp [hunk 7]
```diff
@@ -40,9 +40,1647 @@
...
+ *  @param[in]	step_	-	Increment between each jobs
+ *  @param[in]	max_parallel_	-	Maximum parallel jobs to run
+ *
+ *  @return
+ *  		drmaa2_jarray - returns pointer to drmaa2_jarray if successful
+ *  		NULL and sets last error to DRMAA2_INVALID_ARGUMENT if any argument is invalid
+ *  		NULL and sets last error to DRMAA2_INTERNAL if any internal error occurs
+ */
+
+drmaa2_jarray drmaa2_jsession_run_bulk_jobs(const drmaa2_jsession js_,
```
Comment (arungrover) [inline api/c-binding/drmaa2.cpp]: All parameters in declaration are consts and numeric parameters are long long

File: api/c-binding/drmaa2.cpp (hunk 5 above)
Comment (vinodchitrali) [inline api/c-binding/drmaa2.cpp]: @arun, The declarations are copied from 
https://www.ogf.org/documents/GFD.231.pdf
These interfaces are V2.2 compliant

**Comment by vinodchitrali** [comment]
@arungrover , fixed all comments. Can you check once ?

**Comment by vinodchitrali** [comment]
@mike0042 , Code successfully compiles. One can fork the branch and compile

**Comment by arungrover** [comment]
change looks good!

**Comment by mike0042** [comment]
Until we have tests, I suggest you attach a build log to this and future tickets and check the box for attached test logs. At least that way we have proof that the build didn't break. Thanks!

**Comment by vinodchitrali** [comment]
@mike0042 attached logs to ticket

**Pull request merged**